### PR TITLE
mgr/crash: add process name to crash metadata

### PR DESCRIPTION
**Pull request by trociny**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/global/global_context.h [hunk 1]
```diff
@@ -29,5 +29,6 @@ extern const char *g_assert_condition;
 extern unsigned long long g_assert_thread;
 extern char g_assert_thread_name[4096];
 extern char g_assert_msg[8096];
+extern char g_process_name[4096];
```
Comment (tchaikov) [inline src/global/global_context.h]: the maximum filename length is `NAME_MAX`, normally it is 255. which is much smaller than the maximum length of path. so probably, we can use `NAME_MAX` instead?

File: src/global/global_context.cc [hunk 2]
```diff
@@ -34,3 +34,4 @@ const char *g_assert_condition = 0;
 unsigned long long g_assert_thread = 0;
 char g_assert_thread_name[4096];
 char g_assert_msg[8096];
+char g_process_name[4096];
```
Comment (tchaikov) [inline src/global/global_context.cc]: might want to initialize it to "unknown" or an empty string.

File: src/global/global_context.h (hunk 1 above)
Comment (trociny) [reply src/global/global_context.h]: I just used what was for `g_assert_thread_name`, as they looked similar. But I am fine with `NAME_MAX + 1`.

File: src/global/global_context.cc (hunk 2 above)
Comment (trociny) [reply src/global/global_context.cc]: Static vars are always initialized with zeroes, so it is already an empty string.

**Comment by trociny** [comment]
@tchaikov updated

**Comment by tchaikov** [comment]
@trociny another comment got lost somehow

quote from GETPROGNAME(3bsd)
> In FreeBSD, the name of the program is set by the start-up code that is run before main(); thus, running setprogname() is not necessary.  Programs that desire maximum portability should still call it; on another operating system, these functions may be implemented in a portability library.  Calling setprogname() allows the aforementioned library to learn the program name without modifications to the start-up code.

so to enable other platforms (GNU/Linux for example) to take advantage of this feature, probably we need to call `setprogname()` in `main()` ?

**Comment by trociny** [comment]
> so to enable other platforms (GNU/Linux for example) to take advantage of this feature, probably we need to call setprogname() in main() ?

I am not quite sure what exactly you don't like in the current code. My code just calls `get_process_name` from common. So if here are any compatibility issues they should be resolved there. But as I can see it works well on Linux and FreeBSD right now.

Or don't you like the idea of storing the progname in a static global var on the global context initialization, instead of just calling `get_process_name` (`getprogname`) in the signal handler? I did this way because I was not sure how safe it is to call `get_process_name` in the signal handler, after the crash, on different platforms.

**Comment by tchaikov** [comment]
> But as I can see it works well on Linux and FreeBSD right now.

as per the manpage the progname *might* not be set. my concern was that the progname might not be set. on my linux box, `setprogname()` and `getprogname()` are declared in `bsd/stdlib.h`.  on Debian and its derivative distros, this file is offered by `libbsd-dev`. so i speculated that glibc will not set the `progname` for us on GNU/Linux.

> But as I can see it works well on Linux and FreeBSD right now.

yeah, that's a relief. after digging around, i found that the libbsd implements `getprogname()` like:

https://github.com/freedesktop/libbsd/blob/7cfa2d4530325444fa71c2bdb14e11ba9d4de0b6/src/progname.c#L49-L62

and `program_invocation_short_name` is always around with glibc.

**Comment by trociny** [comment]
> so i speculated that glibc would not set the progname for us on GNU/Linux.

We (`get_process_name`) don't use `getprogname` on linux, but `prctl`.

**Comment by tchaikov** [comment]
> or don't you you like the idea of storing the progname in a static global var on the global context initialization, instead of just calling `get_process_name` (`getprogname`) in the signal handler?

after reading the implementation of libbsd and FreeBSD'c libc, i think it's okay to call `getprogname` in sig handler or it is as bad as storing it beforehand. 

BTW, seems quite a few applications are calling `setprogname()` in `main()`, see https://github.com/lattera/freebsd/search?q=setprogname&type=

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-11-28_00:59:04-rados-wip-kefu2-testing-2018-11-27-2345-distro-basic-smithi/

failures tracked by / caused by
- http://tracker.ceph.com/issues/37432
- http://tracker.ceph.com/issues/20798
- http://tracker.ceph.com/issues/37391
- #19454

**Pull request merged**